### ISSUE TITLE
Problems with character encoding

### ISSUE THREAD

**Issue by EmilJunker** [icatproject/icat.oaipmh#1]
Non-ASCII characters in strings read from the `run.properties` file or received from ICAT's REST interface get scrambled. They should be UTF-8 encoded, but they get interpreted as Latin-1 for some reason, which leads to problems like this:

The string "Helmholtz-Zentum Berlin für Materialien und Energie"
becomes "Helmholtz-Zentum Berlin fÃ¼r Materialien und Energie"

**Comment by EmilJunker** [comment]
I have now investigated this issue and these are my findings.

There are actually two separate issues: One issue affects string from the `run.properties` file, and the other issue affects strings received from ICAT's REST interface by `icat.client`. Both issues have different causes; however, they both have the same symptoms. That is at some point the strings which are meant to be UTF-8 encoded get interpreted as Latin-1 (ISO-8859-1), leading to scrambled characters in the Java String objects.

### Possible workaround
There is actually a way to fix both of these issues, though it is more of a (really bad) workaround and shouldn't be used in production. It is possible to reverse the misinterpreted characters by first decoding the String using ISO-8859-1 by calling `getBytes("ISO-8859-1")` and then passing the resulting bytes to `new String(byte[] bytes)`.

```java
String latin1 = "Helmholtz-Zentum Berlin fÃ¼r Materialien und Energie";  // faulty String
String utf8 = new String(latin1.getBytes("ISO-8859-1"));
logger.info(utf8);  // "Helmholtz-Zentum Berlin für Materialien und Energie"
```

However, this code can [easily break and isn't very efficient](https://stackoverflow.com/questions/46054288/change-utf-8-character-to-latin1-java/46054382#46054382). Also, when a faulty String (like `latin1` in the example above) appears in production, damage has already been done and it is a bad idea to simply reverse it and ignore the cause of the issue.

Hence, I will now explain for each issue what caused it and how it can be fixed.

### Issue with strings from `run.properties`
Let's assume the `run.properties` file of `icat.oaipmh` contains the following line:

```properties
repositoryName = Helmholtz-Zentum Berlin für Materialien und Energie
```

`icat.oaipmh` will then read the string from the `run.properties` file like this.

```java
CheckedProperties props = new CheckedProperties();
props.loadFromResource("run.properties");
String repositoryName = props.getString("repositoryName");
```

This relies on the `CheckedProperties` class from `icat.utils` which in turn is based on the `java.util.Properties` class. To read in the `run.properties` file, `CheckedProperties` [makes a call](https://github.com/icatproject/icat.utils/blob/8ed704cc4e35ca6693eb77ca8d1855d3e78dd1e7/src/main/java/org/icatproject/utils/CheckedProperties.java#L70) to the `load(InputStream inStream)` method from the `java.util.Properties` class.

The `load(InputStream inStream)` method is [described here as follows](https://docs.oracle.com/javase/8/docs/api/java/util/Properties.html#load-java.io.InputStream-):

> Reads a property list (key and element pairs) from the input byte stream. The input stream is in a simple line-oriented format as specified in load(Reader) and **is assumed to use the ISO 8859-1 character encoding**; that is each byte is one Latin1 character.

In other words, all lines from the `run.properties` file get interpreted as ISO-8859-1 instead of UTF-8. This leads to non-ASCII characters getting scrambled.

The easiest way to prevent this is to use Unicode escapes for non-ASCII characters in the `run.properties` file. For example, the character "ü" [may be escaped](https://unicode-table.com/de/00FC/) as "\u00fc". Thus, the line ...

```properties
repositoryName = Helmholtz-Zentum Berlin für Materialien und Energie
```

... would become ...

```properties
repositoryName = Helmholtz-Zentum Berlin f\u00fcr Materialien und Energie
```

I have tested this and it works fine. I will update the documentation accordingly to help others facing similar configuration issues.

**Tl;dr:** This issue is caused by a call to the `load(InputStream inStream)` method from the `java.util.Properties` class which interprets all characters as ISO-8859-1. To fix this, non-ASCII characters must be represented using Unicode escapes, e.g. "\u00fc" for the character "ü".

### Issue with strings received by `icat.client`
`icat.oaipmh` makes queries to ICAT [by calling](https://github.com/icatproject/icat.oaipmh/blob/7fadcfb0d7687611f7743fca778423d15df9e118/src/main/java/org/icatproject/icat_oaipmh/ResponseBuilder.java#L90) the `search(String query)` funtion of the `icat.client` component.

In turn, `icat.client` [makes a call](https://github.com/icatproject/icat.client/blob/2b5d028c896c15c215697ad693826a626945c90c/src/main/java/org/icatproject/icat/client/ICAT.java#L490) to a method called `getString(CloseableHttpResponse response)` when returning the response it received from ICAT's REST interface.

Finally, the `getString(CloseableHttpResponse response)` method [calls](https://github.com/icatproject/icat.client/blob/master/src/main/java/org/icatproject/icat/client/ICAT.java#L319) the `toString(HttpEntity entity)` method from the `EntitiyUtils` class to convert the response to a string.

```java
return EntityUtils.toString(entity);
```

The `toString(HttpEntity entity)` method is [described here](https://hc.apache.org/httpcomponents-core-ga/httpcore/apidocs/org/apache/http/util/EntityUtils.html#toString(org.apache.http.HttpEntity)). By default, it falls back to using the ISO-8859-1 character encoding.

> Read the contents of an entity and return it as a String. The content is converted using the character set from the entity (if any), **failing that, "ISO-8859-1" is used**.

I have found that this is indeed the cause of the issue. I have also found an easy way how this could be fixed. There is an overloaded version of the `toString(HttpEntity entity)` method that takes an additional parameter for the encoding: the `toString(HttpEntity entity, Charset defaultCharset)` method as [described here](https://hc.apache.org/httpcomponents-core-ga/httpcore/apidocs/org/apache/http/util/EntityUtils.html#toString(org.apache.http.HttpEntity,%20java.nio.charset.Charset)).

> Get the entity content as a String, **using the provided default character set** if none is found in the entity.

Thus, the issue could be fixed by changing the above method call to something like this:

```java
return EntityUtils.toString(entity, Charset.defaultCharset());
```

This way, the system's default encoding will be used when stringifying the response which means non-ASCII characters get encoded correctly (at least in my setup that is the case).

I have opened a pull request for `icat.client` to get this resolved: icatproject/icat.client#16

**Tl;dr:** This issue is caused by `icat.client` calling the `toString(HttpEntity entity)` method from the `EntitiyUtils` class which uses the ISO-8859-1 encoding by default. This can be fixed by passing an additional parameter to the method so it uses the system's default encoding instead.

**Comment by RKrahl** [comment]
I can confirm that the suggested fixes work: I built `icat.client` with your changes in icatproject/icat.client#16 and used the result to build `icat.oaipmh`.  And I also applied the encoding of non-ASCII characters in `run.properties` as suggested above.  I didn't see any issues with non-ASCII characters after that.

@stuartpullinger, could you merge the pull request from @EmilJunker in `icat.client` and make a bugfix release?  I'd suggest to require this new `icat.client` version in the `pom.xml` then and to close this issue.

**Comment by EmilJunker** [comment]
I have successfully built `icat.oaipmh` requiring the `icat.client` 4.10.1-SNAPSHOT, and I can confirm that the encoding issues are no longer present after that.

@stuartpullinger, could you make a proper release for `icat.client` 4.10.1 now? After that, we can merge #6 with an updated dependency and close this issue.
